lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -4077,35 +4123,114 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
     auto &original = getOriginal();
     auto &adjoint = getAdjoint();
     auto adjLoc = getAdjoint().getLocation();
-    LLVM_DEBUG(getADDebugStream() << "Running AdjointGen on\n" << original);
-
-    // Create adjoint blocks and arguments.
-    {
-      PostDominanceOrder postDomOrder(&*original.findReturnBB(), postDomInfo);
-      while (auto *origBB = postDomOrder.getNext()) {
-        auto *adjointBB = adjoint.createBasicBlock();
-        adjointBBMap.insert({origBB, adjointBB});
-        postDomOrder.pushChildren(origBB);
-        // If adjoint block is the adjoint entry, continue.
-        if (adjointBB->isEntry())
-          continue;
-        // Otherwise, add a pullback struct argument to the adjoint block.
-        // TODO: Add bb arguments for adjoint values of active values in the
-        // original dominator block. Propagate adjoint values in adjoint bb
-        // terminators below.
-        auto pbStructLoweredType =
-            getPullbackInfo().getPullbackStructLoweredType(origBB);
-        adjointBB->createPhiArgument(
+    LLVM_DEBUG(getADDebugStream() << "Running AdjointEmitter on\n" << original);
+
+    auto *adjGenEnv = getAdjoint().getGenericEnvironment();
+    auto adjGenSig = adjGenEnv
+        ? adjGenEnv->getGenericSignature()->getCanonicalSignature()
+        : nullptr;
+    Lowering::GenericContextScope genericContextScope(
+        getContext().getTypeConverter(), adjGenSig);
+
+    // Get dominated active values in original blocks.
+    // Adjoint values of dominated active values are passed as adjoint block
+    // arguments.
+    DominanceOrder domOrder(original.getEntryBlock(), domInfo);
+    while (auto *bb = domOrder.getNext()) {
+      auto &activeBBValues = activeValues[bb];
+      // If the current block has an immediate dominator, append the immediate
+      // dominator block's active values to the current block's active values.
+      auto *domNode = domInfo->getNode(bb)->getIDom();
+      if (domNode) {
+        auto &activeDomBBValues = activeValues[domNode->getBlock()];
+        activeBBValues.append(activeDomBBValues.begin(),
+                              activeDomBBValues.end());
+      }
+      SmallPtrSet<SILValue, 8> registered(activeBBValues.begin(),
```

Comment:
How about a more common name: "visited"?

```suggestion
      SmallPtrSet<SILValue, 8> visited(activeBBValues.begin(),
```